src/proposed/0180-fpu-switching.md

```diff
@@ -0,0 +1,378 @@
+<!--
+  SPDX-License-Identifier: CC-BY-SA-4.0
+  Copyright 2024 Proofcraft Pty Ltd
+-->
+
+# FPU Context Switching
+
+- Author: Gerwin Klein, Rafal Kolanski, Indan Zupancic
+- Proposed: 2024-06-21
+
+## Summary
+
+We propose to let user level configure policy on when FPU context is switched
+instead of relying on runtime faults to switch FPU context lazily. Policy is
+configured via a flag in the TCB, authorised by the corresponding TCB
+capability.
+
+This improves performance, moves policy from kernel to user, enables verification
+of FPU handling, and fixes an information leakage in current FPU context
+switching.
+
+## Motivation
+
+Current FPU context switching in seL4 disables the FPU on kernel exit, waits for
+a thread to fault when it accesses the FPU, transparently handles the fault,
+saves/restores FPU context, and then returns to the thread.
+
+The motivation for this scheme was that threads that never use the FPU do not pay
+for FPU context switching cost. However, it has the following problems:
+
+- this is a fixed context switching policy in the kernel;
+
+- most of this complexity is invisible to the current formal verification, and
+  therefore not covered by the proofs;
+
+- when switching between domains, this scheme introduces information flow
+  leakage, because the kernel performs actions on behalf of domain A while
+  running in domain B: assuming A and B are the only domains, the first thread
+  running in domain B can determine whether a thread in domain A used the FPU
+  simply by using the FPU itself, since kernel FPU save/restore actions will
+  only occur if another thread used the FPU.
```

Comment:
Would there still be leakage with the current scheme on domain switches if the FPU was always reset back to an empty state on every domain switch?

Reply:
No, if we prove an invariant that the current FPU owner is a thread in the current domain then that will be fine, provided we pull the rest of the FPU switching into the verified subset... but the way it's being done now is not very verification friendly.

Reply:
So should we already want to make a change like what's described in (https://github.com/seL4/seL4/pull/1271) that resets the FPU state on a domain switch?

Reply:
Roughly speaking yes, except that `switchFpuOwner` is not in the verified subset due to it only being previously present on the FPU fault path and paved over during X64 verification, so we'd have to have a magical machine op for it.

Aside: given that the proof abstracts over `numDomains`, it would be preferable that you check with `numDomains > 1` in an `if` and not preprocessor (see comment in `domainConstants` enum).

Reply:
Agreed, that change would break verification in an unpleasant way right now. One of the motivations for doing it properly via this RFC.

Reply:
Can you explain why it's currently considered unverifiable? The proposed scheme also requires the FPU context to be reset on a domain switch and requires tracking the current FPU owner. The behavior for handling this seems independent from how the FPU owner is set, (ie via explicit flag in TCB + context switch, or via use and fpu fault)?

Reply:
It's not entirely unverifiable, just ugly and work intensive. The function `switchFpuOwner` is currently underneath the the machine interface limit -- mentioning it means we'd have to change the machine interface so that it becomes visible, which then draws in a lot of the rest of the FPU handling, which then makes it apparent in verification that there is currently a kernel entry and exit that is not treated. It would be possible, but it'd be a whole lot of work that we'd then need to throw away when we change away from fault-based lazy FPU switching. If we decide that we do want to keep the fault-based lazy switching, that's probably what we'd need to do.